### PR TITLE
Kafka 2257

### PR DESCRIPTION
**Pull request by granders**
This is a 'fake' pull request created solely for the purpose of getting an initial review on this branch before I do the official pull request against the apache/kafka repo.

Hence, this will _not_ be merged into the KAFKA-2276 branch.

@ewencp @junrao 
This patch is intended to replace the existing replication test suite, and follows the proposal for replication test suite outlined here: https://cwiki.apache.org/confluence/display/KAFKA/Roadmap+-+port+existing+system+tests

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java [hunk 1]
```diff
@@ -124,6 +132,23 @@ private static ArgumentParser argParser() {
                 .choices(0, 1, -1)
                 .metavar("ACKS")
                 .help("Acks required on each produced message. See Kafka docs on request.required.acks for details.");
+        
+        parser.addArgument("--producer-prop")
```
Comment (ewencp) [inline clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java]: Other tools (at least producer performance, possibly others) just allow a list of `prop=value` arguments. Any reason not to be consistent with those (or whatever the KIP-4 discussion might have ended up agreeing upon)?

File: clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java [hunk 2]
```diff
@@ -124,6 +132,23 @@ private static ArgumentParser argParser() {
...
+                .required(false)
+                .type(String.class)
+                .dest("userSpecifiedProducerProps")
+                .metavar("PRODUCER-PROP")
+                .help("Any custom producer properties of the form key=val.");
+
+        parser.addArgument("--close-timeout")
+                .action(store())
+                .required(false)
+                .setDefault(10L)
```
Comment (ewencp) [inline clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java]: Curious about the default here. Should it be significantly higher? For this tool, it's pretty important to report as accurate information as possible. I'm concerned that if you happen to do something like stop this producer when you're also bouncing brokers, are there cases where this default might not be sufficient and cause tests to appear to fail (e.g. it will actually succeed, slowly, but because of this timeout we log it as failed).

Perhaps an important related question is what types of situations might we get stuck waiting for a long time (in a system test, not a real production environment) such that we wouldn't want to make this a very conservatively large number? This may also be relevant with the up coming max.block.ms setting from KIP-19 (where I assume the default will be ok, but possibly for a tool like this we'd want to be even more conservative than the default).

File: clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java [hunk 3]
```diff
@@ -124,6 +132,23 @@ private static ArgumentParser argParser() {
...
+                .type(Long.class)
+                .metavar("CLOSE-TIMEOUT")
+                .dest("closeTimeoutSeconds")
+                .help("When SIGTERM is caught, wait at most this many seconds for unsent messages to flush before stopping the VerifiableProducer process.");
```
Comment (ewencp) [inline clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java]: Generally configs for anything time related are in ms. I think there may be an exception or two when the time scale is expected to be much larger (log rolling, log retention have flexible units). Not sure if we want to try to be as consistent as possible (use ms) or use the units that most likely match what the user will want to express (use seconds).

File: clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java [hunk 4]
```diff
@@ -150,8 +174,24 @@ public static VerifiableProducer createFromArgs(String[] args) {
             producerProps.put(ProducerConfig.ACKS_CONFIG, Integer.toString(res.getInt("acks")));
             // No producer retries
             producerProps.put("retries", "0");
```
Comment (ewencp) [inline clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java]: Unrelated but I just saw this -- why isn't this just ProducerConfig.RETRIES_CONFIG?

File: clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java [hunk 5]
```diff
@@ -150,8 +174,24 @@ public static VerifiableProducer createFromArgs(String[] args) {
...
+            
+            // Properties specified using the --producer-prop key=val format override
+            // other properties
+            List<String> userSpecifiedProducerProps = res.getList("userSpecifiedProducerProps");
+            for (String prop: userSpecifiedProducerProps) {
```
Comment (ewencp) [inline clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java]: Do we have code for this somewhere already. Looks like it should be reusable across a bunch of tools.

File: clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java [hunk 6]
```diff
@@ -193,7 +233,7 @@ String errorString(Exception e, String key, String value, Long nowMs) {
     
         JSONObject obj = new JSONObject();
         obj.put("class", this.getClass().toString());
-        obj.put("name", "producer_send_error");
+        obj.put("name", "f_error");
```
Comment (ewencp) [inline clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java]: Why this change? Looks like it might be a typo?

File: clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java [hunk 7]
```diff
@@ -233,6 +273,11 @@ String successString(RecordMetadata recordMetadata, String key, String value, Lo
     
         public void onCompletion(RecordMetadata recordMetadata, Exception e) {
             synchronized (System.out) {
+                if (VerifiableProducer.this.stopProducing.get()) {
```
Comment (ewencp) [inline clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java]: Not sure about this behavior. The whole point of this producer is to log complete information about what it sent and what it knows about the success of those sends. If we ignore some callbacks, we won't have logged all the data. It seems very likely we would regularly see "extra" data at consumers if we used this approach.

One solution would be to still log these events, but adjust the output when you see an error and `stopProducing` is true, e.g. use `producer_send_closing_error` instead of `producer_send_error` to indicate it may have just been due to a timeout (although again, with the addition of KIP-19's patch with client-side timeouts, we may have to rethink what exactly these messages mean since sends could succeed but timeout too quickly).

File: tests/kafkatest/services/kafka.py [hunk 8]
```diff
@@ -54,6 +54,14 @@ def start(self):
                 topic_cfg["topic"] = topic
                 self.create_topic(topic_cfg)
 
+        # Sanity check - are the broker processes actually running?
```
Comment (ewencp) [inline tests/kafkatest/services/kafka.py]: Just a side note, but it'd be really nice to have some utility code to help parallelize stuff like this.

File: tests/kafkatest/services/kafka.py [hunk 9]
```diff
@@ -199,29 +208,53 @@ def restart_node(self, node, wait_sec=0, clean_shutdown=True):
     def leader(self, topic, partition=0):
         """ Get the leader replica for the given topic and partition.
         """
-        cmd = "/opt/kafka/bin/kafka-run-class.sh kafka.tools.ZooKeeperMainWrapper -server %s " \
-              % self.zk.connect_setting()
-        cmd += "get /brokers/topics/%s/partitions/%d/state" % (topic, partition)
-        self.logger.debug(cmd)
+        topic_partition_data = self.zk_get_data("/brokers/topics/%s/partitions/%d/state" % (topic, partition))
```
Comment (ewencp) [inline tests/kafkatest/services/kafka.py]: These paths don't account for ZK chroots. Or is that just handled by the zkconnect setting?

File: tests/kafkatest/services/kafka.py [hunk 10]
```diff
@@ -199,29 +208,53 @@ def restart_node(self, node, wait_sec=0, clean_shutdown=True):
...
+
+        if topic_partition_data is None:
+            raise Exception("Error finding data for topic %s and partition %d." % (topic, partition))
+        self.logger.info(topic_partition_data)
```
Comment (ewencp) [inline tests/kafkatest/services/kafka.py]: Logging this to info for every lookup seems like overkill.

File: tests/kafkatest/services/kafka.py [hunk 11]
```diff
@@ -199,29 +208,53 @@ def restart_node(self, node, wait_sec=0, clean_shutdown=True):
...
+
+        leader_idx = int(topic_partition_data["leader"])
+        self.logger.info("Leader for topic %s and partition %d is now: %d" % (topic, partition, leader_idx))
```
Comment (ewencp) [inline tests/kafkatest/services/kafka.py]: Same here actually -- if it's important enough to log, the caller of this method should probably log it. Logging in accessors gets annoying if anything ends up needing to call it frequently.

File: tests/kafkatest/services/kafka.py [hunk 12]
```diff
@@ -199,29 +208,53 @@ def restart_node(self, node, wait_sec=0, clean_shutdown=True):
...
 
+        controller_idx = int(controller_data["brokerid"])
+        self.logger.info("Controller is: %d" % controller_idx)
+        controller_node = self.get_node(controller_idx)
+
+        if controller_node is None:
+            raise Exception("Found no node corresponding to the id: %d" % controller_idx)
+        return controller_node
+
+    def zk_get_data(self, path):
```
Comment (ewencp) [inline tests/kafkatest/services/kafka.py]: Should we just move this to the Zookeeper service? Seems like it's a better home for it.

File: tests/kafkatest/services/kafka.py [hunk 13]
```diff
@@ -199,29 +208,53 @@ def restart_node(self, node, wait_sec=0, clean_shutdown=True):
...
+        """
+        Get data from zookeeper on the given path. This method assumes the data is in JSON format.
+
+        :param path Zookeeper path to query for data
+        :param node optional node on which query will be run
```
Comment (ewencp) [inline tests/kafkatest/services/kafka.py]: docs seem out of date, there's no `node` parameter here

File: tests/kafkatest/services/verifiable_producer.py [hunk 14]
```diff
@@ -23,16 +23,18 @@ class VerifiableProducer(BackgroundThreadService):
     logs = {
         "producer_log": {
             "path": "/mnt/producer.log",
-            "collect_default": False}
+            "collect_default": True}
```
Comment (ewencp) [inline tests/kafkatest/services/verifiable_producer.py]: Why did this change? Isn't there a data size problem since it logs every single message?

File: tests/kafkatest/services/verifiable_producer.py [hunk 15]
```diff
@@ -63,6 +65,11 @@ def start_cmd(self):
             cmd += " --max-messages %s" % str(self.max_messages)
         if self.throughput > 0:
             cmd += " --throughput %s" % str(self.throughput)
+        cmd += " --close-timeout %s" % str(self.close_timeout_sec)
+
+        if self.producer_props is not None:
```
Comment (ewencp) [inline tests/kafkatest/services/verifiable_producer.py]: Could change the default to be an empty list (or zero-tuple if you prefer making it immutable), which avoids `is not None` checks like this littering the code.

File: tests/kafkatest/tests/replication_test.py [hunk 16]
```diff
@@ -15,45 +15,86 @@
...
 from kafkatest.services.zookeeper import ZookeeperService
 from kafkatest.services.kafka import KafkaService
 from kafkatest.services.verifiable_producer import VerifiableProducer
 from kafkatest.services.console_consumer import ConsoleConsumer
 
-import signal
 import time
 
 
+"""
```
Comment (ewencp) [inline tests/kafkatest/tests/replication_test.py]: This is sort of a docstring but a) not associated with anything and b) doesn't give any context.

I think you either want to move this up to be the module's docstring (top of the file) or move this into the class's docstring. The class's docstring also provides some context.

File: tests/kafkatest/tests/replication_test.py [hunk 17]
```diff
@@ -15,45 +15,86 @@
...
+3 brokers, 3 topics, 3 partitions, 3 replicas on each topic
+min.insync.replicas == 2
+ack -1
+failures: [leader, follower, controller] X [clean, hard, soft]
+One test cases with with ack 1
+One test case with compression toggled on
+
+"""
+
+# Failure types
```
Comment (ewencp) [inline tests/kafkatest/tests/replication_test.py]: Seems like we could just split these into two variables:

1. Type of death (clean, hard, soft)
2. Permanence (bounce vs kill).

This would clean up some of the if statements in the code farther down in this file.

File: tests/kafkatest/tests/replication_test.py [hunk 18]
```diff
@@ -15,45 +15,86 @@
...
     """Replication tests.
     These tests verify that replication provides simple durability guarantees by checking that data acked by
     brokers is still available for consumption in the face of various failure scenarios."""
 
     def __init__(self, test_context):
-        """:type test_context: ducktape.tests.test.TestContext"""
         super(ReplicationTest, self).__init__(test_context=test_context)
 
-        self.topic = "test_topic"
+        self.replication_factor = 3
```
Comment (ewencp) [inline tests/kafkatest/tests/replication_test.py]: weird mix where we store this as a field, but num partitions and min insync replicas are not?

File: tests/kafkatest/tests/replication_test.py [hunk 19]
```diff
@@ -15,45 +15,86 @@
...
         self.zk = ZookeeperService(test_context, num_nodes=1)
-        self.kafka = KafkaService(test_context, num_nodes=3, zk=self.zk, topics={self.topic: {
-                                                                    "partitions": 3,
-                                                                    "replication-factor": 3,
-                                                                    "min.insync.replicas": 2}
-                                                                })
-        self.producer_throughput = 10000
-        self.num_producers = 1
-        self.num_consumers = 1
+        self.topics = {"topic1": {
```
Comment (ewencp) [inline tests/kafkatest/tests/replication_test.py]: Not a big deal, but these are very redundant, could just point all of them to the same dict.

File: tests/kafkatest/tests/replication_test.py [hunk 20]
```diff
@@ -15,45 +15,86 @@
...
+                       "topic2": {
+                            "partitions": 3,
+                            "replication-factor": self.replication_factor,
+                            "min.insync.replicas": 2},
+                       "topic3": {
+                            "partitions": 3,
+                            "replication-factor": self.replication_factor,
+                            "min.insync.replicas": 2}
+                    }
+        self.topic_names = ["topic1", "topic2", "topic3"]
```
Comment (ewencp) [inline tests/kafkatest/tests/replication_test.py]: Not sure I understand the purpose of having these topics. If we only fail brokers for one of them, what are the other ones accomplishing? They may end up as collateral damage, but what information does that give us since we aren't logging info about how those topics would be affected? Or is this holdover from the previous system test design?

File: tests/kafkatest/tests/replication_test.py [hunk 21]
```diff
@@ -15,45 +15,86 @@
...
+        self.topic_to_fail = "topic1"  # We'll induce failures on this topic
+
+        self.kafka = KafkaService(test_context, num_nodes=3, zk=self.zk, topics=self.topics)
+        self.producer_throughput = 100000
+        self.num_producers = 3
```
Comment (ewencp) [inline tests/kafkatest/tests/replication_test.py]: things that are constants like this can be nice to set as class variables and use all-caps if they really are intended as constants. It groups all the key configs together at the top of the class.

File: tests/kafkatest/tests/replication_test.py [hunk 22]
```diff
@@ -15,45 +15,86 @@
...
+        self.num_consumers = 3
+
+        self.producers = [VerifiableProducer(self.test_context, 1, self.kafka, topic, throughput=self.producer_throughput)
```
Comment (ewencp) [inline tests/kafkatest/tests/replication_test.py]: Did we not write these services to support multiple nodes? Maybe it would make more sense to do that? I'm not sure yet how common this pattern will be, but if we need to instantiate multiple concurrent producers of the same type, using a single service seems simpler (and can also centralize the logic we have below for starting all the nodes, compiling all the lists of acked/unacked messages together, etc).

File: tests/kafkatest/tests/replication_test.py [hunk 23]
```diff
@@ -75,28 +116,35 @@ def run_with_failure(self, failure):
...
-        if not wait_until(lambda: self.producer.num_acked > 5, timeout_sec=5):
-            raise RuntimeError("Producer failed to start in a reasonable amount of time.")
-        failure()
-        self.producer.stop()
+        self.logger.debug("Producing messages...")
+        self.start_producers()
+
+        self.logger.debug("Driving failures...")
+        self.drive_failures(failure, node_type)
+        time.sleep(5)  # Keep on producing for a few more seconds
```
Comment (ewencp) [inline tests/kafkatest/tests/replication_test.py]: Why is this necessary? Do we not wait long enough during `drive_failures` to make sure we're back in a good state?

Would be nice to keep the arbitrary `sleep()` calls to a minimum :)

File: tests/kafkatest/tests/replication_test.py [hunk 24]
```diff
@@ -75,28 +116,35 @@ def run_with_failure(self, failure):
...
+        self.stop_producers()
 
-        self.acked = self.producer.acked
-        self.not_acked = self.producer.not_acked
-        self.logger.info("num not acked: %d" % self.producer.num_not_acked)
-        self.logger.info("num acked:     %d" % self.producer.num_acked)
+        self.acked = [producer.acked for producer in self.producers]
+        self.not_acked = [producer.not_acked for producer in self.producers]
+        self.logger.info("num not acked on topic %s: %d" % (producer.topic, producer.num_not_acked))
```
Comment (ewencp) [inline tests/kafkatest/tests/replication_test.py]: I think we're only logging this info for the last producer because `producer` is still bound to the var from the list comprehension?

File: tests/kafkatest/tests/replication_test.py [hunk 25]
```diff
@@ -104,44 +152,84 @@ def run_with_failure(self, failure):
...
+    def drive_failures(self, failure, node_type):
+
+        if failure in {SOFT_KILL, HARD_KILL, CLEAN_KILL}:
+            self.kill(failure, node_type)
+        elif failure in {SOFT_BOUNCE, HARD_BOUNCE, CLEAN_BOUNCE}:
+            self.bounce(failure, node_type, self.num_bounce)
+        else:
+            raise RuntimeError("Invalid failure type")
+
+    def bounce(self, failure, node_type, num_bounce):
```
Comment (ewencp) [inline tests/kafkatest/tests/replication_test.py]: `bounce` and `kill` seem like they are functionality of the service? These are probably going to be reused across a number of tests, so they should probably just start out as part of the service. Same with the associated constants (which are actually even more general and we might want to move to ducktape once we have better process management integration).

File: tests/kafkatest/tests/replication_test.py [hunk 26]
```diff
@@ -104,44 +152,84 @@ def run_with_failure(self, failure):
...
+        for i in range(num_bounce):
+            node_to_signal = self.fetch_broker_node(self.topic_to_fail, 0, node_type)
+
+            if failure == CLEAN_BOUNCE or failure == HARD_BOUNCE:
+                self.kafka.restart_node(node_to_signal, wait_sec=5, clean_shutdown=(failure == CLEAN_BOUNCE))
+            elif failure == SOFT_BOUNCE:
+                self.kafka.signal_node(node_to_signal, "SIGSTOP")
+                time.sleep(20)
```
Comment (ewencp) [inline tests/kafkatest/tests/replication_test.py]: So many magic sleep constants :) I realize we don't have a great solution for this at the moment -- would something simple added to KafkaService that can check log files for relevant messages help to avoid these? Or maybe even something in ZK we can check that would indicate the broker is back online and ready?

**Comment by ewencp** [comment]
@granders This patch is smaller than I thought it was going to be and much of it is for utilities that can be reused in other tests -- this is a great sign that we're on the right track! I left a bunch of comments but haven't actually tested yet, so take everything with a grain of salt :)

File: clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java (hunk 6 above)
Comment (granders) [inline clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java]: Yeah this must have been a typo. Reverted.

File: clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java (hunk 5 above)
Comment (granders) [inline clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java]: Nothing like this in clients/tools, since this is the first introduction of an option parsing library,

There may be something analogous in core.

However I'm not sure if it's good to add a dependency on core, where jopt-simple is used. Also, as noted with the problems parsing "-1", jopt-simple had some impressively obnoxious corner cases.

File: clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java (hunk 2 above)
Comment (granders) [inline clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java]: Yeah I'm not sure the best solution here.

Basically every replication test results in waiting for a very long time after the initial trigger of the close() method, and typically the wait can be *much* longer than all of the other test logic.

One alternative method might be the throttle the throughput more aggressively.

File: clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java (hunk 3 above)
Comment (granders) [inline clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java]: Changed to ms

File: tests/kafkatest/services/kafka.py (hunk 10 above)
Comment (granders) [inline tests/kafkatest/services/kafka.py]: This is pretty useful to have at some level - changed to debug

File: tests/kafkatest/services/kafka.py (hunk 9 above)
Comment (granders) [inline tests/kafkatest/services/kafka.py]: Hm, is this really necessary? There is no mechanism in ZookeeperService for creating a chroot

File: tests/kafkatest/services/kafka.py (hunk 11 above)
Comment (granders) [inline tests/kafkatest/services/kafka.py]: moved to test code

File: tests/kafkatest/services/kafka.py (hunk 13 above)
Comment (granders) [inline tests/kafkatest/services/kafka.py]: updated

File: tests/kafkatest/services/kafka.py (hunk 12 above)
Comment (granders) [inline tests/kafkatest/services/kafka.py]: yup, moved

File: tests/kafkatest/services/verifiable_producer.py (hunk 14 above)
Comment (granders) [inline tests/kafkatest/services/verifiable_producer.py]: Changed while debugging. Reverted.

File: tests/kafkatest/services/verifiable_producer.py (hunk 15 above)
Comment (granders) [inline tests/kafkatest/services/verifiable_producer.py]: Hm. I suppose it's ok to have a mutable default argument in the __init__ method since it will only be called once in the object's lifetime.

Changed to empty dict.

File: tests/kafkatest/tests/replication_test.py (hunk 16 above)
Comment (granders) [inline tests/kafkatest/tests/replication_test.py]: Yeah, this was primarily a "note to self". Deleted.

File: tests/kafkatest/tests/replication_test.py (hunk 18 above)
Comment (granders) [inline tests/kafkatest/tests/replication_test.py]: gone

File: tests/kafkatest/tests/replication_test.py (hunk 19 above)
Comment (granders) [inline tests/kafkatest/tests/replication_test.py]: updated

File: clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java (hunk 5 above)
Comment (ewencp) [inline clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java]: Right, I just thought a few tools allowed a set of generic config opts in opt=value form already, but maybe those are all housed in core. Maybe we need a few more tools here before we can figure out which things should be refactored into utilities.

File: clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java (hunk 2 above)
Comment (ewencp) [inline clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java]: Why are they waiting that long? Do they just have a ton of data buffered? Shouldn't the default buffer size keep the outstanding data to a reasonable size so `close()` wouldn't have to wait for very long?

File: tests/kafkatest/services/kafka.py (hunk 10 above)
Comment (ewencp) [inline tests/kafkatest/services/kafka.py]: With stuff like this, I like to ask whether anyone might ever call this in a pretty tight loop, e.g. in order to wait to meet some condition (like leader failover) but detect it fairly promptly. In those cases, logging in an accessor can be dangerous (huge amount of log output).

File: tests/kafkatest/services/kafka.py (hunk 9 above)
Comment (ewencp) [inline tests/kafkatest/services/kafka.py]: That's part of why I was asking about the zkconnect setting. It's just something to keep in mind because we're increasing the number of ways we're relying on the semantics of the ZookeeperService class, and I realized that chroots are something we may want to test at some point. May not require anything in this patch, could be a parameter where we only use the default, or something else. It's worth thinking through the implications as we add more utility methods like this.

File: clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java (hunk 1 above)
Comment (granders) [inline clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java]: The only place I kinda see this is in kafka-topic.sh, but neither KIP-4 nor KIP-14 make reference to allowing a list of `prop=value`.

Currently kafka-topic.sh doesn't really have this behavior (should it?...):
This works (one --config per config):
```
kafka_fork $ bin/kafka-topics.sh --topic abc --zookeeper=localhost:2181 --create --partitions 1 --replication-factor 1 --config max.message.bytes=111 --config flush.ms=222
```

This silently ignores the second config (flush.ms):
```
kafka_fork $ bin/kafka-topics.sh --topic abc --zookeeper=localhost:2181 --create --partitions 1 --replication-factor 1 --config max.message.bytes=111 flush.ms=222
```

From a quick scan of https://reviews.apache.org/r/34554/diff/7#3 it looks like this behavior is unchanged in the KIP-4 patches.

That all said, I changed the name to "--config" to be consistent.

File: clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java (hunk 2 above)
Comment (granders) [inline clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java]: My current hypothesis is that printing to stdout in each callback has enough overhead that a huge lag develops between the stuff queued up for stdout and the messages actually sent.

If you run the producer with no close timeout at with default throughput of -1, and then hit ctrl-c, you'll see many many thousands of messages printed after the fact, and wait a very long time for everything to flush. I don't think the buffer size is the issue (~16,000 by default), since many many than 16,000 messages print after hitting ctrl-c.

If you comment out the println statements in the callback and replace with some computations, the producer exits almost immediately when you hit ctrl-c.

File: clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java (hunk 4 above)
Comment (granders) [inline clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java]: oops, good call. Updated

File: clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java (hunk 2 above)
Comment (ewencp) [inline clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java]: How much more backlogged can it get than the buffer size? Aren't those callbacks invoked in the producer's network thread such that they would block any further sends until they complete? 

In any case, this probably means we need to reduce the cost of this logging. It probably needs some profiling to determine what the issue is for sure, but some obvious things that might be affecting how expensive this is:

* `System.out` will flush with every newline, i.e. for every call to println. This could add up since each line isn't that long.
* There's a lot of redundant work being done, e.g. creating JSONObject, getting the class name, etc.
* How fast is this JSON library? It could end up being faster to just do the encoding manually even though that's less than ideal.
* We're converting to a string, then writing to `System.out` rather than writing directly.

Also, be careful evaluating this by running directly in the console. If the issue has anything to do with the output, involving a console can significantly affect performance. You're better off logging something so you can tell when Ctrl-C was hit, then using the subsequent log entries and their timestamps to see how many more events there were and how long they took. Along the same lines, is this log being written locally and then synced via ducktape or is it running via SSH and with the SSH output being processed on the test runner? The SSH connection could also end up being a bottleneck.

File: tests/kafkatest/tests/replication_test.py (hunk 20 above)
Comment (granders) [inline tests/kafkatest/tests/replication_test.py]: This is the way Jun and I sketched the design for the port, but that doesn't mean that past Jun and Geoff were right.

The verification logic is actually checking every topic, so we are exercising failover in the presence of multiple topics. Maybe that is useful enough to justify the small added complexity (and run time), but I'm not sure.

Note that I dropped FOLLOWER failover, since with 3 partitions (even if we had 1 topic) we're guaranteed to be hitting a follower for some partition some of the time.

File: tests/kafkatest/tests/replication_test.py (hunk 23 above)
Comment (granders) [inline tests/kafkatest/tests/replication_test.py]: removed

File: tests/kafkatest/tests/replication_test.py (hunk 24 above)
Comment (granders) [inline tests/kafkatest/tests/replication_test.py]: removed

File: tests/kafkatest/tests/replication_test.py (hunk 26 above)
Comment (granders) [inline tests/kafkatest/tests/replication_test.py]: This was a janky experiment to trigger reelection so that the previous leader could come back online after the new one was elected.

Replaced this with a more precise `wait_until(...)` call

File: tests/kafkatest/tests/replication_test.py (hunk 20 above)
Comment (ewencp) [inline tests/kafkatest/tests/replication_test.py]: I'm not worried about a little more complexity, but it isn't clear just from reading the code what's going on so it probably at least needs some comment. Also, are we actually *guaranteed* one of them is going to be a follower some of the time, or is it possible we get unlucky with leadership and always manage to kill the leaders of all topics?

These tests require pretty deep understanding of how Kafka works and sometimes rely on assumptions that could change over time (e.g., how partition assignment works, guarantees about distribution of partition leadership, etc). We have some good info lower down in this file about the details about how we do the verification. With tests like these it'd also be good to have methodology info in the class docstring.

**Comment by granders** [comment]
Done with this local PR
